src/dbnode/storage/bootstrap/bootstrapper/base.go

```diff
@@ -76,133 +74,261 @@ func (b baseBootstrapper) String() string {
+			// Calculate bootstrap time ranges.
+			indexRequired = namespace.IndexRunOptions.ShardTimeRanges.Copy()
+			indexCurrRequested = currNamespace.IndexRunOptions.ShardTimeRanges.Copy()
+			indexCurrFulfilled = indexCurrRequested.Copy()
+			indexCurrFulfilled.Subtract(currResult.IndexResult.Unfulfilled())
+
+			indexUnfulfilled := indexRequired.Copy()
+			indexUnfulfilled.Subtract(indexCurrFulfilled)
+
+			// Modify the unfulfilled result.
+			currResult.IndexResult.SetUnfulfilled(indexUnfulfilled.Copy())
 
-	fulfilledRanges := result.ShardTimeRanges{}
-	fulfilledRanges.AddRanges(currStatus.fulfilled)
-	fulfilledRanges.AddRanges(nextStatus.fulfilled)
-	unfulfilled := totalRanges.Copy()
-	unfulfilled.Subtract(fulfilledRanges)
+			// Set the next bootstrapper required ranges.
+			nextNamespace.IndexRunOptions.ShardTimeRanges = indexUnfulfilled.Copy()
+		} else {
+			// NB(r): Make sure to always set an empty requested range so IsEmpty
+			// does not cause nil ptr deref.
+			nextNamespace.IndexRunOptions.ShardTimeRanges = result.ShardTimeRanges{}
+		}
+
+		// Set the modified result.
+		currResults.Results.Set(id, currResult)
+
+		// Set the next bootstrapper namespace run options if we need to bootstrap
+		// further time ranges.
+		if !nextNamespace.DataRunOptions.ShardTimeRanges.IsEmpty() ||
+			!nextNamespace.IndexRunOptions.ShardTimeRanges.IsEmpty() {
+			next.Namespaces.Set(id, nextNamespace)
+		}
 
-	step.mergeResults(unfulfilled)
+		// Log the result.
+		successLogFields := append(logFieldsCopy(logFields), []zapcore.Field{
+			zap.String("namespace", id.String()),
+			zap.Int("numShards", len(currNamespace.Shards)),
+		}...)
+		successLogFields = append(successLogFields,
+			zap.Duration("dataRangeRequested", dataCurrRequested.MinMaxRange()))
+		successLogFields = append(successLogFields,
+			zap.Duration("dataRangeFulfilled", dataCurrFulfilled.MinMaxRange()))
+		successLogFields = append(successLogFields,
+			zap.Duration("indexRangeRequested", indexCurrRequested.MinMaxRange()))
+		successLogFields = append(successLogFields,
+			zap.Duration("indexRangeFulfilled", indexCurrFulfilled.MinMaxRange()))
+
+		successLogFields = append(successLogFields,
+			zap.Int("dataNumSeries", currResult.DataMetadata.NumSeries))
+		if namespace.Metadata.Options().IndexOptions().Enabled() {
+			successLogFields = append(successLogFields,
+				zap.Int("indexNumSeries", currResult.IndexMetadata.NumSeries))
+		}
+
+		b.log.Info("bootstrapping from source completed successfully",
+			successLogFields...)
 
-	unattemptedNextRanges := currRanges.Copy()
-	unattemptedNextRanges.Subtract(currStatus.fulfilled)
-	if !nextAttempted {
-		// If we have never attempted the next time range then we want to also
-		// attempt the ranges we didn't even try to attempt
-		unattemptedNextRanges.AddRanges(nextRanges)
 	}
 
+	// Unless next bootstrapper is required, this is the final results.
+	finalResults := currResults
```

Comment:
meganit: Is `finalResults` necessary?

Reply:
I wanted to make it clear we move on to actually modifying the curr results in place from the next results.  And updating `currResults` from `nextResult` seemed strange?